In [1]:
import scipy.stats as st
import numpy as np
import pandas as pd


In [2]:
male = [40,44,53,57]
female = [60,54,46,41]
high_school = [60,40]
bachelors = [54,44]
masters = [46,53]
phd = [41,57]
educations = ['high_school','bachelors','masters','phd','high_school','bachelors','masters','phd']
sex = ['M','M','M','M','F','F','F','F']
marks = male+female
marks

[40, 44, 53, 57, 60, 54, 46, 41]

In [3]:
df = pd.DataFrame({'Education':educations, 'Marks': marks,'Sex':sex})
df

Education  Marks Sex
0  high_school     40   M
1    bachelors     44   M
2      masters     53   M
3          phd     57   M
4  high_school     60   F
5    bachelors     54   F
6      masters     46   F
7          phd     41   F

In [4]:
crossTab = pd.crosstab([df.Sex,df.Marks],df.Education,margins =True)
crossTab

Education  bachelors  high_school  masters  phd  All
Sex Marks                                           
F   41             0            0        0    1    1
    46             0            0        1    0    1
    54             1            0        0    0    1
    60             0            1        0    0    1
M   40             0            1        0    0    1
    44             1            0        0    0    1
    53             0            0        1    0    1
    57             0            0        0    1    1
All                2            2        2    2    8

In [5]:
df1 = pd.crosstab(df.Sex,df.Education,df.Marks,aggfunc='sum',margins= True)
df1

Education  bachelors  high_school  masters  phd  All
Sex                                                 
F                 54           60       46   41  201
M                 44           40       53   57  194
All               98          100       99   98  395

In [6]:
df1.columns = ['bachelors','high_school','masters','phd','gender_total']
df1.index  =['Female','Male','Combined']
df1

bachelors  high_school  masters  phd  gender_total
Female           54           60       46   41           201
Male             44           40       53   57           194
Combined         98          100       99   98           395

In [7]:
df2 = df1.iloc[0:2,0:4]
df2

bachelors  high_school  masters  phd
Female         54           60       46   41
Male           44           40       53   57

In [8]:
'''for a test of independence, we use the same chi sqred formula that we used for the\
goodness of fit test.The main difference is we have to calculate the expected count of each cell in 2d'''

'for a test of independence, we use the same chi sqred formula that we used for thegoodness of fit test.The main difference is we have to calculate the expected count of each cell in 2d'

In [9]:
df3 = np.outer(df1['gender_total'][0:2],df1.loc['Combined'][0:4])/395.0
df3 = pd.DataFrame(df3)
df3.columns = ['high_school','bachelors','masters','phd']
df3.index = ['Female','Male']
df3

high_school  bachelors    masters        phd
Female    49.868354  50.886076  50.377215  49.868354
Male      48.131646  49.113924  48.622785  48.131646

In [10]:
x = df3-df2
print(x)
chi_sqrd_stat = (x**2/df3).sum().sum()
print(chi_sqrd_stat)

        bachelors  high_school   masters       phd
Female  -3.113924   -10.131646  4.377215  8.868354
Male     5.113924     8.131646 -4.377215 -8.868354
8.14077142853548


In [11]:
cvalue = st.chi2.ppf(q = 0.95,df = 3)#dof
print(cvalue)

7.814727903251179


In [12]:
p_val = 1 - st.chi2.cdf(x= chi_sqrd_stat,df =3)
print(p_val)

0.043190233390898114


In [16]:
result = st.chi2_contingency(observed= df2)
print(result)
#chi2_contigency is used to conduct of test of independece automatically given a frequency table of observed table


(8.006066246262538, 0.045886500891747214, 3, array([[49.86835443, 50.88607595, 50.37721519, 49.86835443],
       [48.13164557, 49.11392405, 48.62278481, 48.13164557]]))


here output chi_sqr statistic 8,pvalue is 0.045 and dof  =3
critical value with dof = 3... is 7.815<8(chisqr value)
therefore we reject null hypothesis
conclusion-- education level dependes on gender with 5% level of significance

In [13]:

grp1 = [51, 45, 33, 45, 67]
grp2  =[23, 43, 23, 43, 45]
grp3 = [56, 76, 74, 87, 56]
statistics,pvalue = st.f_oneway(grp1,grp2,grp3)
print('f statistic value {},pvalue {}'.format(statistics,pvalue))
if pvalue < 0.05:
    print('True')
else:
    print('False')
print('-------------------')
print('the test result suggests the grps have different same sample means in this\
      \nexample since p value is significant @ 99% confidence\n  here p value returned is 0.00305 less than 0.05')

f statistic value 9.747205503009463,pvalue 0.0030597541434430556
True
-------------------
the test result suggests the grps have different same sample means in this      
example since p value is significant @ 99% confidence
  here p value returned is 0.00305 less than 0.05


In [14]:
st.f_oneway([10, 20, 30, 40, 50],[5,10,15, 20, 25])

grp_1 = [10, 20, 30, 40, 50] 
grp_2 = [5,10,15, 20, 25]
mean1=np.mean(grp_1)
mean2 = np.mean(grp_2)
grp_1_submean  = []
grp_2_submean  = []
add1=0
add2 =0
for i in grp_1:
    add1 += (i-mean1)**2
for j in grp_2:
    add2 += (j-mean2)**2
    
var1 = add1/(len(grp_1)-1)
var2 = add2/(len(grp_2)-1)

ftest = var1/var2
print('ftest',ftest)

ftest 4.0
